File: antigravity/archive/run-3-php-v2.0/src/workers/email_worker.php
<?php
require_once __DIR__ . '/../../vendor/autoload.php';

use Tina4\Queue;
use Tina4\Messenger;

// Initialize the environment if needed by starting App container context
// Tina4 autoloads env variables from .env
$queue = new Queue(topic: 'emails');

echo "Email queue worker started. Listening on topic: 'emails'...\n";

foreach ($queue->consume('emails') as $job) {
    $payload = $job->payload;
    echo "[" . date("c") . "] Processing job ID: {$job->id} to {$payload['to']}...\n";

    try {
        $mailer = new Messenger();
        $htmlBody = "
            <html>
            <body style='font-family: Arial, sans-serif; padding: 20px;'>
                <h2>Lend Library - Borrowing Receipt</h2>
                <p>Hello <strong>{$payload['name']}</strong>,</p>
                <p>You have successfully borrowed: <strong>{$payload['title']}</strong>.</p>
                <p><strong>Borrow Date:</strong> {$payload['borrow_date']}</p>
                <p><strong>Due Date:</strong> <span style='color: #d9534f; font-weight: bold;'>{$payload['due_date']}</span></p>
                <p>Please return the book by the due date. Thank you!</p>
            </body>
            </html>
        ";

        $textBody = "Lend Library - Borrowing Receipt\n\n" .
                    "Hello {$payload['name']},\n\n" .
                    "You have successfully borrowed: {$payload['title']}.\n" .
                    "Borrow Date: {$payload['borrow_date']}\n" .
                    "Due Date: {$payload['due_date']}\n\n" .
                    "Please return the book by the due date. Thank you!";

        $result = $mailer->send(
            $payload['to'],
            "Book Borrowed: {$payload['title']}",
            $htmlBody,
            ["text_body" => $textBody]
        );

        if ($result["success"]) {
            echo "  Email sent successfully!\n";
            $job->complete();
        } else {
            echo "  Failed sending email: {$result['error']}\n";
            $job->fail($result['error']);
        }
    } catch (\Throwable $e) {
        echo "  Exception: {$e->getMessage()}\n";
        $job->fail($e->getMessage());
    }
}

File: antigravity/a-vanilla/run-2/src/services/email_worker.py
import time
from tina4_python.queue import Queue
from tina4_python.debug import Log
from tina4_python.messenger import Messenger

class EmailWorker:
    def __init__(self):
        self.queue = Queue(topic="emails", max_retries=3)
        self._running = False

    def __call__(self, ctx):
        self.run(ctx)

    def run(self, ctx):
        self._running = True
        Log.info("Email worker started")

        while self._running and not ctx.stop_event.is_set():
            job = self.queue.pop()

            if job is None:
                ctx.stop_event.wait(1)
                continue

            try:
                self._send_email(job.payload)
                job.complete()
                Log.info(f"Email sent successfully to {job.payload.get('to')}")
            except Exception as exc:
                job.fail(str(exc))
                Log.warning(f"Email failed to {job.payload.get('to')}: {exc}")

    def stop(self):
        self._running = False
        Log.info("Email worker stopped")

    def _send_email(self, payload):
        Log.info(f"Sending email receipt to {payload['to']} for book {payload['title']} due on {payload['due_date']}")
        
        # Instantiate Messenger to send the email
        mailer = Messenger()
        mailer.send(
            to=payload["to"],
            subject=payload["subject"],
            body=payload["body"]
        )

